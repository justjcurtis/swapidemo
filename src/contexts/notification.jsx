import React, { useState, useRef } from 'react';

const NotificationContext = React.createContext();
const NotificationProvider = ({ children }) => {
  const [notifications, setNotifications] = useState([]);
  const timeout = useRef(null);

  const notify = (message, type) => {
    const created = Date.now();
    setNotifications([...notifications, { message, type, created }]);
    if (timeout.current) clearTimeout(timeout.current)
    timeout.current = setTimeout(() => {
      setNotifications((notifications) => notifications.filter((n) => n.created < Date.now() - 3000));
      timeout.current = null;
    }, 3000);
  };

  return (
    <NotificationContext.Provider value={{ notifications, notify }}>
      {children}
    </NotificationContext.Provider>
  );
}

export { NotificationProvider, NotificationContext }
