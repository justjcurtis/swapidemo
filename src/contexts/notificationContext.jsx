import React, { useState, useRef } from 'react';
import { NOTIFICATION_TIMEOUT } from '../constants/notificationContants';

const NotificationContext = React.createContext();
const NotificationProvider = ({ children }) => {
  const [notifications, setNotifications] = useState([]);
  const timeout = useRef(null);

  const notify = (message, type) => {
    const created = Date.now();
    setNotifications([...notifications, { message, type, created }]);
    if (timeout.current) clearTimeout(timeout.current)
    timeout.current = setTimeout(() => {
      setNotifications((notifications) => notifications.filter((n) => n.created < Date.now() - NOTIFICATION_TIMEOUT));
      timeout.current = null;
    }, NOTIFICATION_TIMEOUT);
  };

  return (
    <NotificationContext.Provider value={{ notifications, notify }}>
      {children}
    </NotificationContext.Provider>
  );
}

export { NotificationProvider, NotificationContext }
