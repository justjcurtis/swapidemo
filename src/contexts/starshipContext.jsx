import React, { useContext, useState } from 'react';
import { NotificationContext } from './notificationContext';

const StarshipContext = React.createContext();
const StarshipProvider = ({ children }) => {
  const { notify } = useContext(NotificationContext)
  const [isLoading, setLoading] = useState(true);

  const fetchStarships = async (pageNumber) => {
    try {
      setLoading(true);
      const url = `https://swapi.dev/api/starships/?page=${pageNumber ?? 1}`
      const response = await fetch(url);
      const data = await response.json();
      return data;
    } catch (error) {
      notify('Failed to fetch starships, please try again later', 'error');
      console.error(error);
    } finally {
      setLoading(false);
    }
  };

  return (
    <StarshipContext.Provider value={{ fetchStarships, isLoading }}>
      {children}
    </StarshipContext.Provider>
  );
}

export { StarshipProvider, StarshipContext }
