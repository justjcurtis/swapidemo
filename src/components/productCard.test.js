import { render, waitFor, act } from '@testing-library/react'
import ProductCard from './productCard'
import starships from '../testData/starships.json'
import { NotificationContext, NotificationProvider } from '../contexts/notificationContext'

describe("ProductCard", () => {
    test("renders correctly", () => {
        const product = starships[0]
        const { container } = render(
            <NotificationProvider>
                <ProductCard product={product} />
            </NotificationProvider>
        )
        expect(container.innerHTML).toMatchSnapshot()
    })
    describe("when amount is 0", () => {
        test("Calls 'notify' with error when buy button is clicked", async () => {
            const product = starships[0]
            const notify = jest.fn()
            const { getByText } = render(
                <NotificationContext.Provider value={{ notify }}>
                    <ProductCard product={product} />
                </NotificationContext.Provider>
            )
            act(() => {
                getByText("BUY").click()
            })
            await waitFor(() => {
                expect(notify).toHaveBeenCalledWith("Please add at least 1 item to your basket", "error")
            })
        })
    })

    describe("when amount is greater than 0", () => {
        test("Calls 'notify' with success when buy button is clicked", async () => {
            const product = starships[0]
            const notify = jest.fn()
            const { getByText, rerender } = render(
                <NotificationContext.Provider value={{ notify }}>
                    <ProductCard product={product} />
                </NotificationContext.Provider>
            )
            act(() => {
                getByText("+").click()
            })
            rerender(
                <NotificationContext.Provider value={{ notify }}>
                    <ProductCard product={product} />
                </NotificationContext.Provider>
            )
            act(() => {
                getByText("BUY (1)").click()
            })
            await waitFor(() => {
                expect(notify).toHaveBeenCalledWith("1 CR90 corvette added to basket (â‚¹3500000)", "success")
            })
        })
    })
})

