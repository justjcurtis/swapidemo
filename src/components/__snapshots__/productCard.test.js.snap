// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard renders correctly 1`] = `"<div style="display: flex; flex-direction: column; justify-content: center; align-items: center; padding: 0px 0.5rem; background: rgb(68, 68, 68); border-radius: 5px; box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1); margin: 2rem;"><h1 style="color: rgb(255, 255, 255); font-size: 2rem; font-weight: bold;">CR90 corvette</h1><div style="width: 100%; margin: 0px; display: flex; flex-direction: row; justify-content: space-between; padding: 0.5rem; background: rgb(51, 51, 51); background-color: rgb(102, 102, 102);"><span style="text-align: left; font-weight: bold;">Class: </span><span style="text-align: right;">Corvette</span></div><div style="width: 100%; margin: 0px; display: flex; flex-direction: row; justify-content: space-between; padding: 0.5rem; background: rgb(51, 51, 51); background-color: rgb(51, 51, 51);"><span style="text-align: left; font-weight: bold;">Manufacturer: </span><span style="text-align: right;">Corellian Engineering Corporation</span></div><div style="width: 100%; margin: 0px; display: flex; flex-direction: row; justify-content: space-between; padding: 0.5rem; background: rgb(51, 51, 51); background-color: rgb(102, 102, 102);"><span style="text-align: left; font-weight: bold;">Model: </span><span style="text-align: right;">CR90 Corvette</span></div><div style="width: 100%; margin: 0px; display: flex; flex-direction: row; justify-content: space-between; padding: 0.5rem; background: rgb(51, 51, 51); background-color: rgb(51, 51, 51);"><span style="text-align: left; font-weight: bold;">Cargo Capacity: </span><span style="text-align: right;">3,000,000 kg</span></div><div style="width: 100%; margin: 0px; display: flex; flex-direction: row; justify-content: space-between; padding: 0.5rem; background: rgb(51, 51, 51); background-color: rgb(102, 102, 102);"><span style="text-align: left; font-weight: bold;">Passengers: </span><span style="text-align: right;">600</span></div><div style="width: 100%; margin: 0px; display: flex; flex-direction: row; justify-content: space-between; padding: 0.5rem; background: rgb(51, 51, 51); background-color: rgb(51, 51, 51);"><span style="text-align: left; font-weight: bold;">Crew: </span><span style="text-align: right;">30</span></div><div style="width: 100%; margin: 0px; display: flex; flex-direction: row; justify-content: space-between; padding: 0.5rem; background: rgb(51, 51, 51); background-color: rgb(102, 102, 102);"><span style="text-align: left; font-weight: bold;">Consumables: </span><span style="text-align: right;">1 years</span></div><div style="width: 100%; margin: 0px; display: flex; flex-direction: row; justify-content: space-between; padding: 0.5rem; background: rgb(51, 51, 51); background-color: rgb(51, 51, 51);"><span style="text-align: left; font-weight: bold;">Megalights: </span><span style="text-align: right;">60</span></div><div style="width: 100%; margin: 0px; display: flex; flex-direction: row; justify-content: space-between; padding: 0.5rem; background: rgb(51, 51, 51); background-color: rgb(102, 102, 102);"><span style="text-align: left; font-weight: bold;">Hyperdrive Rating: </span><span style="text-align: right;">2</span></div><div style="width: 100%; margin: 0px; display: flex; flex-direction: row; justify-content: space-between; padding: 0.5rem; background: rgb(51, 51, 51); background-color: rgb(51, 51, 51);"><span style="text-align: left; font-weight: bold;">Max Atmosphering Speed: </span><span style="text-align: right;">950 km/h</span></div><div style="width: 100%; margin: 0px; display: flex; flex-direction: row; justify-content: space-between; padding: 0.5rem; background: rgb(51, 51, 51); background-color: rgb(102, 102, 102);"><span style="text-align: left; font-weight: bold;">Length: </span><span style="text-align: right;">150 m</span></div><div style="width: 100%; margin: 0px; display: flex; flex-direction: row; justify-content: space-between; padding: 0.5rem; background: rgb(51, 51, 51); background-color: rgb(51, 51, 51);"><span style="text-align: left; font-weight: bold;">Cost: </span><span style="text-align: right;">â‚¹3,500,000</span></div><div style="margin: 1rem 0px; display: flex; flex-direction: row; justify-content: space-evenly; width: 100%;"><button class="cds--btn cds--btn--primary" type="button">-</button><button class="cds--btn cds--btn--primary" type="button">BUY</button><button class="cds--btn cds--btn--primary" type="button">+</button></div></div>"`;
